src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,177 @@
+
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> current = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(current.prev, value, current);
+        current.prev = (current.prev != null) ? current.prev.next
+                = newNode : (head = newNode);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            node = new Node<>(null, list.get(i), null);
+            tail.next = node;
+            node.prev = tail;
+            tail = node;
+            size++;
+        }
+
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> current = findNodeByIndex(index);
+        return current.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        check(index);
+        node = new Node(value);
+        Node<T> current = findNodeByIndex(index);
+        if (index == 0) {
+            current.next.prev = node;
+            node.prev = null;
+            node.next = current.next;
+            head = node;
+        } else {
+            node.prev = current.prev;
+            node.next = current.next;
+            current.next.prev = node;
+            current.prev.next = node;
+        }
+        T oldItem = current.item;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        check(index);
+        Node<T> current = findNodeByIndex(index);
+        if (current.equals(head)) {
+            head = head.next;
+        } else {
+            current.prev.next = current.next;
+        }
+        if (current.equals(tail)) {
+            tail = tail.prev;
+        } else {
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        boolean found = false;
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.item == object || current.item != null && current.item.equals(object)) {
+                if (current == head) {
+                    head = current.next;
+                    if (head != null) {
+                        head.prev = null;
+                    }
+                } else if (current == tail) {
+                    tail = current.prev;
+                    tail.next = null;
+                } else {
+                    current.prev.next = current.next;
+                    current.next.prev = current.prev;
+                }
+                found = true;
+                size--;
+                break;
+            }
+            current = current.next;
+        }
+        return found;
```

Comment:
You should create a separate `unlink(Node node)` method to handle the removal logic, as this code is repeated in the `remove(int index)` method.